Expose toggle to textfield's opacity animation.

blquinn opened the pull request:
This exposes the toggle to cupertino/material textfield's opacity animation so that users can disable the cursor opacity animation if they choose to.

Sort of fixes https://github.com/flutter/flutter/issues/59327

*List which issues are fixed by this PR. You must list at least one issue.*

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

blquinn committed:
Add opacity animation toggle to CupertinoTextField.

blquinn committed:
Add opacity animation toggle to Material TextField.

blquinn committed:
Merge branch 'master' into disable-textfield-opacity-animation

jmagman commented:
Missing documentation for the new params:
```
dartdoc:stdout: Generating docs for library cupertino from package:flutter/cupertino.dart...
dartdoc:stderr:   error: undefined macro [flutter.widgets.editableText.cursorOpacityAnimates]
dartdoc:stderr:     from cupertino.CupertinoTextField.cursorOpacityAnimates: (file:///tmp/flutter%20sdk/packages/flutter/lib/src/cupertino/text_field.dart:666:14)
...
dartdoc:stderr:   error: undefined macro [flutter.widgets.editableText.cursorOpacityAnimates]
dartdoc:stderr:     from material.TextField.cursorOpacityAnimates: (file:///tmp/flutter%20sdk/packages/flutter/lib/src/material/text_field.dart:586:15)
```

blquinn committed:
Fix docs

blquinn committed:
Merge branch 'disable-textfield-opacity-animation' of github.com:Blquinn/flutter into disable-textfield-opacity-animation

blquinn committed:
Merge branch 'master' into disable-textfield-opacity-animation

Blquinn commented:
@jmagman I've updated the docs. Not sure why that macro didn't work, but it's building now.

gnprice commented:
The macro wasn't working because there wasn't a definition for it. If you'd like to use the handy macro feature in the docs, you can add a `@template` marker with the same syntax that's used on the neighboring field:
```dart
  /// {@template flutter.widgets.editableText.cursorRadius}
  /// How rounded the corners of the cursor should be.
  ///
  /// By default, the cursor has no radius.
  /// {@endtemplate}
  final Radius? cursorRadius;

  /// Whether the cursor will animate from fully transparent to fully opaque
  /// during each cursor blink.
  ///
  /// By default, the cursor opacity will animate on iOS platforms and will not
  /// animate on Android platforms.
  final bool cursorOpacityAnimates;
```

That `{@template` etc. syntax is what defines the meaning of `{@macro flutter.widgets.editableText.cursorRadius}`, and you can do the same thing to make it work the same way for `cursorOpacityAnimates`.

Blquinn commented:
@gnprice Oh nice, thanks for the tip!

justinmc commented on packages/flutter/lib/src/material/text_field.dart:
```diff
@@ -581,6 +582,13 @@ class TextField extends StatefulWidget {
+  /// during each cursor blink.
+  ///
+  /// By default, the cursor opacity will animate on iOS platforms and will not
+  /// animate on Android platforms.
```
Can you use a macro/template here instead of duplicating this?  I see some discussion above about a macro not working before but let me know if you can't get it.

justinmc reviewed (commented):
I'm on board with this.  Since cursorOpacityAnimates already exists on EditableText, I see no reason why we shouldn't expose it on TextField and CupertinoTextField.

Can you add tests for this?  I think you can simply add one for each of TextField and CupertinoTextField, and verify that underlying EditableText has the correct cursorOpacityAnimates value.

Also, have you verified that turning this off does indeed fix the CPU performance issue?  I see you commented on the issue with some hard numbers, were those taken from disabling and enabling this on EditableText?

blquinn committed:
Use macro for cursorOpacityAnimates docs.

blquinn committed:
Merge branch 'disable-textfield-opacity-animation' of github.com:Blquinn/flutter into disable-textfield-opacity-animation

blquinn committed:
Add tests for cursorOpacityAnimates

Blquinn commented:
@justinmc I've added the macro for the docs and added tests you suggested for material and cupertino. I have also manually verified that the toggle overrides the default platform value correctly on a linux computer and mac.

> Also, have you verified that turning this off does indeed fix the CPU performance issue? I see you commented on the issue with some hard numbers, were those taken from disabling and enabling this on EditableText?

Correct, I tested this by disabling/enabling the parameter in the text field examples. Nothing but the text field is present in the test app.

I have checked the CPU usage on multiple platforms (linux, macOS, iOS, windows) on physical devices, not emulators and in release mode. Each platform shows the same trend -- with the animation enabled I see a constant 10-16% cpu usage and with it disabled I see 0% cpu usage.

On desktop platforms I measured CPU using htop and the systems system monitor program. For iOS I used xCode's profiler/monitor thing.

blquinn committed:
Fix lint in test

blquinn committed:
More linting issues.

blquinn committed:
Merge branch 'master' into disable-textfield-opacity-animation

blquinn committed:
Merge branch 'master' into disable-textfield-opacity-animation

justinmc reviewed (approved):
LGTM :+1:.  Thank you for the fixes!

Blquinn commented:
@justinmc Sweet, thanks for the review! What exactly is the process for getting this merged?

werainkhatri reviewed (approved):
congrats on your first contribution!

![blue flutter](https://user-images.githubusercontent.com/44755140/226993764-1176007f-8d3f-49c5-9474-9820da471f13.png)

justinmc commented:
Sorry @Blquinn, the process was just to find a secondary reviewer, which @werainkhatri has been nice enough to provide.  Now this will merge automatically when the tree is green.

Blquinn commented:
Awesome; no problem at all. Thank you both!

gnprice commented:
The tree is green, but the "Google testing" check for this PR seems to have gotten stuck:
* #123289

One workaround is to update the PR so the check gets restarted. I'm not sure if there are other workarounds a Googler could apply.

blquinn committed:
Merge branch 'master' into disable-textfield-opacity-animation
